### PR TITLE
Active user cruncher with MMQ, HTTP and multiple intervals.

### PR DESCRIPTION
**Pull request by grixa**
CC @dkorolev @mzhurovich

### CODE REVIEW

**Comment by dkorolev** [comment]
```

clang++-3.6 -std=c++11 -Wall -W -ftemplate-backtrace-limit=0  -ftemplate-depth=10000 -DCURRENT_CI -o ".current/test" "test.cc" -pthread -ldl
In file included from test.cc:1511:
In file included from ./crunchers/test.cc:38:
./crunchers/cruncher_active_users.h:130:38: error: C++ requires a type specifier
      for all declarations
                               const buffer_size = 1024*1024)
                               ~~~~~ ^
```

:-(

File: crunchers/cruncher_active_users.h [hunk 1]
```diff
@@ -90,12 +91,115 @@ struct ActiveUsersCruncherImpl {
   user_map_t users_map_;
   std::chrono::microseconds current_us_;
   std::chrono::microseconds interval_;
-  uint64_t events_seen_;
 };
 
 template <typename NAMESPACE>
 using ActiveUsersCruncher = CTFO::StreamCruncher<ActiveUsersCruncherImpl<NAMESPACE>>;
 
+CURRENT_STRUCT(ResponseGetActiveUsersShort) {
```
Comment (grixa) [inline crunchers/cruncher_active_users.h:99]: Is there any use of this "short" response?

File: crunchers/cruncher_active_users.cc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  ParseDFlags(&argc, &argv);
+
+  if (FLAGS_helpconfig) {
+    current::reflection::StructSchema schema;
+    schema.AddType<ActiveUsersCruncherParams>();
+    std::cout << schema.GetSchemaInfo().Describe<current::reflection::Language::FSharp>(false);
+  } else {
+    ActiveUsersCruncherParams params;
+    if (FLAGS_defaultconfig) {
+      params.intervals = {std::chrono::seconds(10), std::chrono::hours(1), std::chrono::hours(24)};
```
Comment (dkorolev) [inline crunchers/cruncher_active_users.cc]: Why not have these in `ActiveUsersCruncherParams`'s default constructor?

File: crunchers/cruncher_active_users.h [hunk 3]
```diff
@@ -90,12 +91,115 @@ struct ActiveUsersCruncherImpl {
...
+  struct Message final {
+    Message() = default;
+    Message(const entry_t& e, idxts_t idxts) : idxts(idxts), event(new entry_t(e)), type(Event) {}
+    Message(entry_t&& e, idxts_t idxts) : idxts(idxts), event(new entry_t(std::move(e))), type(Event) {}
+    explicit Message(Request&& r) : request(new Request(std::move(r))), type(GetData) {}
+    idxts_t idxts;
+    std::unique_ptr<entry_t> event;
+    std::unique_ptr<Request> request;
+    enum { Event, GetData } type;
+  };
```
Comment (dkorolev) [inline crunchers/cruncher_active_users.h:120]: When initializing a `unique_ptr`, `new` is always best to be replaced by `make_unique`.

Design suggestion: Why not have `Message` a base class of virtual functions only, introduce `EventMessage` and `RequestMessage`, and have the MMQ operate with `std::unique_ptr<Message>`-s?

Publishing events to the MMQ would be as simple as `.Publish(std::make_unique<{Event|Request}Message>(...))`.

And both Request and `entry_t` have their own move constructors, so you won't need to use the dreaded pointers within messages then.

File: crunchers/cruncher_active_users.h [hunk 4]
```diff
@@ -90,12 +91,115 @@ struct ActiveUsersCruncherImpl {
...
+struct ActiveUsersMultiCruncherImpl {
+  using cruncher_t = ActiveUsersCruncherImpl<NAMESPACE>;
+  using entry_t = typename NAMESPACE::CTFOLogEntry;
+
+  struct Message final {
+    Message() = default;
+    Message(const entry_t& e, idxts_t idxts) : idxts(idxts), event(new entry_t(e)), type(Event) {}
+    Message(entry_t&& e, idxts_t idxts) : idxts(idxts), event(new entry_t(std::move(e))), type(Event) {}
+    explicit Message(Request&& r) : request(new Request(std::move(r))), type(GetData) {}
+    idxts_t idxts;
```
Comment (dkorolev) [inline crunchers/cruncher_active_users.h]: `const` ?

File: crunchers/cruncher_active_users.h [hunk 5]
```diff
@@ -90,12 +91,115 @@ struct ActiveUsersCruncherImpl {
...
+  using mmq_t = current::mmq::MMQ<Message, IntermediateSubscriber<Message>, 1024 * 1024>;
+
+  ActiveUsersMultiCruncherImpl(const std::vector<std::chrono::microseconds>& intervals,
+                               uint16_t port,
+                               const std::string& route,
+                               const buffer_size = 1024*1024)
+      : port_(port),
+        last_event_us_(std::chrono::microseconds(0)),
+        mmq_subscriber_([this](Message&& message, idxts_t) {
+          switch (message.type) {
```
Comment (dkorolev) [inline crunchers/cruncher_active_users.h]: These would be cleaner if implemented via virtual functions.

File: crunchers/cruncher_active_users.h (hunk 3 above)
Comment (grixa) [reply crunchers/cruncher_active_users.h:120]: I thought about this solution. How these virtual functions would look like? For example, they could be two getters, which return `Request` and `entry_t`. But in that case they can't be pure, and they can't return values or references, only pointers. Am I right?

Comment (dkorolev) [reply crunchers/cruncher_active_users.h:120]: These two virtual functions could do the work within themselves, w/o exposing internal details, such as returning pointers or references.

Comment (grixa) [reply crunchers/cruncher_active_users.h:120]: You mean, pass some kind of callback inside? Or move there the code from `OnEventInternal` and `HandleGetDataInternal`?

Comment (dkorolev) [reply crunchers/cruncher_active_users.h:120]: Move the code from `OnEventInternal` and `HandleGetDataInternal` into those virtual functions. If the need contest, passing `(*this)` as the parameter is still clean.

**Comment by dkorolev** [comment]
Commit https://github.com/KnowSheet/CTFO_BE/commit/87bfc47 :+1: :+1: :+1: LGTM!

Good to merge?

**Comment by grixa** [comment]
`this_t` -> `self_t` or just `self` ?

**Comment by dkorolev** [comment]
`this_t` vs. `self_t` is probably a matter of taste, I was more talking about the name of the variable itself: `s/cruncher/self/g`.

**Comment by grixa** [comment]
OK, one minute..

**Comment by dkorolev** [comment]
:-) спасибо! Мёржу как сделаешь?

File: crunchers/cruncher_active_users.h [hunk 6]
```diff
@@ -109,21 +109,21 @@ CURRENT_STRUCT(ResponseGetActiveUsers) {
...
-  using this_t = ActiveUsersMultiCruncherImpl<NAMESPACE>;
+  using self_t = ActiveUsersMultiCruncherImpl<NAMESPACE>;
   using cruncher_t = ActiveUsersCruncherImpl<NAMESPACE>;
   using entry_t = typename NAMESPACE::CTFOLogEntry;
 
   class Message {
    public:
     virtual ~Message() = default;
-    virtual void Handle(this_t& cruncher) = 0;
+    virtual void Handle(self_t& self) = 0;
```
Comment (dkorolev) [inline crunchers/cruncher_active_users.h:119]: :+1:

**Comment by grixa** [comment]
Да)

**Comment by dkorolev** [comment]
Готово :-)

**Pull request merged**